Introduce `Solver` class

ddundo opened the pull request:
Closes #238. Closes #22.

Summary (so far):
- Introduced `goalie.solver.Solver` that all user-defined problems should inherit from
- `MeshSeq` now only has these methods left: `count_elements`, `count_vertices`, `plot`, `set_meshes`
- `get_solver` no longer requires a nested `solver` function (#22)
- Renamed `AdjointMeshSeq` and `GoalOrientedMeshSeq` to `AdjointSolver` and `GoalOritentedSolver`, respectively

<details><summary>Example workflow</summary>
<p>

```
class MySolver(Solver):
    def get_function_spaces(self, mesh):
        ...
        
    def get_solver(self, index):
        # No longer has a nested solver function
        ...
        
    def get_initial_conditions(self):
        ...
        
time_partition = ... # as before
mesh_sequence = MeshSeq(list_of_meshes)
solver = MySolver(time_partition, mesh_sequence)
solver.solve_forward()
solver.fixed_point_iteration(adaptor)
```

</p>
</details> 

____

Won't do in this PR:
- ~~Rename `Solver` and its methods? Maybe `Model` would be better and then rename `get_initial_condition` to `initial_condition`, etc.?~~ Can introduce a `Model` class in another PR. This would require #279 or a similar solution
- ~~Pass `field_names` and `field_types` to `Solver` so that it's self-contained~~

Commit:
#238: Introduce bare-bones Solver class

Commit:
#238: Temp dirty modification of MeshSeqs

Commit:
#238: Temp modify burgers2 demo

ddundo commented:
@acse-ej321 we very briefly discussed this on the meeting and @jwallwork23 is in principle in favour of this. If you are as well, maybe you could try this approach for your enriched mesh seq issue and see if that solves it? I think it should since the two are now separated :)

jwallwork23 commented:
Please change branch target to [develop](https://github.com/mesh-adaptation/goalie/tree/develop).

ddundo committed:
Merge branch 'develop' into 238_solver_class

ddundo committed:
Use duck typing in Solver instead of ABC

ddundo committed:
Expand Solver docstrings WIP [skip ci]

ddundo committed:
Revert temporary changes in mesh_seq.py

ddundo committed:
Strip out everything solver-related from MeshSeq

ddundo committed:
Revert temporary changes in adjoint.py

ddundo committed:
Copy over relevant MeshSeq methods to Solver [skip ci]

ddundo committed:
Update variable names in Solver (WIP)

ddundo committed:
Update demos for testing [skip ci]

ddundo committed:
Introduce FixedPointIterator; move relevant methods from MeshSeq

ddundo committed:
Move _transfer from MeshSeq to Solver

ddundo committed:
Remove get_qoi from Solver

ddundo committed:
Add logging methods to Solver [skip ci]

ddundo committed:
Update MeshSeq tests

ddundo committed:
Update syntax in _outputs_consistent check

ddundo committed:
Remove FixedPointIterator and move back methods

ddundo committed:
Move element count convergence check to Solver

ddundo committed:
Again make get_initial_condition return 0s by default

ddundo committed:
Simplify get_* functions' signature

ddundo committed:
Remove unused check_element_count_convergence method in MeshSeq

ddundo committed:
Add test_solver.py which tests ported fns from MeshSeq

ddundo commented:
I have finished part of this PR, by porting stuff from `mesh_seq.py` to `solver.py` (which now has the `Solver` class). I modified all related tests (all tests in `tests` pass but not in `tests/adjoint`) and I modified three demos: `burgers.py`, `burgers2.py`, and `burgers-hessian.py`. 

@jwallwork23 @acse-ej321 would you mind telling me whether you like the new structure please? Just so I don't waste time updating the rest if you don't like it :)

jwallwork23 commented on goalie/solver.py:
```diff
@@ -0,0 +1,556 @@
+__all__ = ["Solver"]
+
+
+class Solver:
```
I thought we'd decided to call it `Model`?

ddundo replied:
If I understood correctly, @acse-ej321 suggested that `Model` was a new class that would be passed to `Solver`. So something like

```
class BurgersModel(Model):
    def get_solver(...):
        ....
        # etc

solver = goalie.solver.Solver(BurgersModel, mesh_sequence, time_partition)
```

I don't remember whether we agreed to do this in the end?

And I thought it's best to leave it for another PR anyway since that would modify quite a few other things (including the signature of these `get_` functions, as I mentioned in #279) and I didn't want to have a PR with too many things at once for the sake of easier reviewing. But I'm happy to do this all at once if you're happy to review it as such :)

jwallwork23 replied:
Oh I thought it was just that it'd be called Model rather than Solver because it also knows about function spaces, QoIs, etc. To me, the solver is just the bit that solves the PDE (which may or may not involve timestepping). So in this case, that'd be the solver method.

ddundo replied:
This is what I proposed originally (see first comment in this PR) but then Eleda suggested that it could be a separate class. And I don't remember what the conclusion was :)

What would you prefer?

acse-ej321 replied:
Apologies for the confusion. @ddundo interpreted my comments correctly in that it would look something like 
```python
class BurgersModel(Model):
    def get_function_space(...):
    def get_inital_conditions(...):
    def get_qoi(...):
    def get_solver(...):
        ....
        # etc

class Solver(...):
   def solve_forward(...):
   def solve_adjoint(...):
   def indicate_errors(...):
   ...
   # etc
solver = goalie.solver.Solver(BurgersModel, mesh_sequence, time_partition)`
```
I was agreeing with the first comment in the original PR, that  Model seemed a better name rather than Solver. As @jwallwork23 commented, `Model` encompasses better the essence of the problem definition.  The additional suggestion was that maybe the methods associated with solving such as `fixed_point_iteration`, `indicate errors` and `forward_solve` might be separated out as well. In the example above, this would be `Solver` but could even be an independent `FixedPointIteration` class as in [Separate FixedPointIterator class #244](https://github.com/mesh-adaptation/goalie/issues/244). That the definition of the problem could be further decoupled from the PDE solving.

jwallwork23 replied:
Oh okay. In the pseudocode you give there @acse-ej321, the nice thing is that the Model requires user-defined methods but the Solver doesn't, right? You might pass parameters to the Solver, but you don't need to give it code.

acse-ej321 replied:
@jwallwork23 - yes I think that is correct. The user-defined methods are all in the problem/model class and the Solver just executes those methods.

ddundo replied:
The only disadvantage I can think of is that it adds another class that users need to be mindful of. Not sure how much of a concern that is

jwallwork23 replied:
Yes, in support of @acse-ej321's suggested approach.

jwallwork23 commented on goalie/solver.py:
```diff
@@ -0,0 +1,556 @@
+        """
+        Create the :class:`~.FunctionData` instance for holding solution data.
+        """
+        self._solutions = ForwardSolutionData(self.time_partition, self.function_spaces)
```
What about if we want to solve adjoint problems?

ddundo replied:
I haven't fully modified that yet, but I was thinking of essentially just renaming `AdjointMeshSeq` to `AdjointSolver` (subclass of `Solver`). I wouldn't remove or add any new methods.

ddundo replied:
Does that sound good?

jwallwork23 replied:
I don't think it's necessary to have separate AdjointX subclasses any more. I think I mentioned this in the meeting last week. Instead, we can just have a `Solver` that implements both `solve_forward` and `solve_adjoint`.

But let's leave that refactoring for a later PR. It's sufficient to have a separate `AdjointSolver` for now, thanks.

jwallwork23 commented on goalie/solver.py:
```diff
@@ -0,0 +1,556 @@
+        :rtype: :class:`list` of :class:`firedrake.function.Function`\s
+        """
+        solver_kwargs = solver_kwargs or {}
+        N = self.num_subintervals  # FIXME
```
Why FIXME?

ddundo replied:
Sorry, I forgot to delete this comment :)

jwallwork23 commented on goalie/solver.py:
```diff
@@ -0,0 +1,556 @@
+            if update_params is not None:
+                update_params(self.params, self.fp_iteration)
+
+            # Solve the forward problem over all meshes
+            self.solve_forward(solver_kwargs=solver_kwargs)
```
I notice this fixed point iteration only solves the forward problem. What about adjoints? I was more inclined to have a separate class for handling fixed point iteration (#244).

ddundo replied:
In this PR I thought I'd keep the two fixed point iterator methods where they are (in `Solver` and `AdjointSolver`) in order to not have a massive PR. But again, I'm happy to introduce `FixedPointIterator`s in this PR if you don't mind reviewing a massive PR

jwallwork23 replied:
Fair enough. Yes, leave them for now and that can come in the follow-up refactor.

jwallwork23 commented on demos/burgers-hessian.py:
```diff
@@ -19,17 +19,17 @@
 
-def get_function_spaces(mesh):
-    return {"u": VectorFunctionSpace(mesh, "CG", 2)}
+class MySolver(Solver):
```
Would prefer `BurgersSolver` to `MySolver`. Similarly elsewhere.

jwallwork23 commented on demos/burgers-hessian.py:
```diff
@@ -19,17 +19,17 @@
-def get_solver(mesh_seq):
-    def solver(index):
-        u, u_ = mesh_seq.fields["u"]
+    def get_solver(self, index):
```
If it's going to take this different form then we'd probably be better off calling it `solver` rather than `get_solver`?

ddundo replied:
And then similarly `get_initial_condition` -> `initial_condition` etc.?

jwallwork23 replied:
Actually let's leave such name changes for a future PR, to avoid confusing things. But it's worth noting that we're changing the nature of the `get_solver` method here so it might be worth revising how we're naming the methods.

joewallwork reviewed (commented):
Thanks for your work on this, @ddundo. A few thoughts.

ddundo committed:
Refactor `AdjointMeshSeq` -> `AdjointSolver`

ddundo committed:
Refactor `GoalOrientedMeshSeq` -> `GoalOrientedSolver` [skip ci]

jwallwork23 commented:
> Thanks for taking a look @jwallwork23. To demonstrate what I meant in the replies to your comments, I just pushed the refactored `AdjointMeshSeq` and `GoalOrientedMeshSeq` (and you can see it in action with the `burgers_ee.py` demo).

These look good, thanks. The PR is taking great shape!

ddundo committed:
`MeshSeq` no longer accepts a single mesh for multiple subintervals

ddundo committed:
Update mesh_seq.py demo and add a code output block

ddundo committed:
Add `Model` class; update burgers.py [skip ci]

ddundo commented:
@jwallwork23 @acse-ej321 could you please take a look at the small commit https://github.com/mesh-adaptation/goalie/pull/239/commits/73f36107b8a18e43753afc5df3dabc7f238db56f where I added the `Model` class and updated the `burgers.py` demo? Since the `Model` class is now separate from `Solver`, I added `time_partition`, `fields`, `function_spaces`, and `meshes` to the signatures of `get_` functions.

Do you agree with this? I find it a bit messy, but I don't have a different suggestion.

acse-ej321 commented on goalie/__init__.py:
does `from goalie.model import * #noqa` need to be added?

ddundo replied:
It does indeed, thanks! Added it :)

acse-ej321 commented:
> @jwallwork23 @acse-ej321 could you please take a look at the small commit [73f3610](https://github.com/mesh-adaptation/goalie/commit/73f36107b8a18e43753afc5df3dabc7f238db56f) where I added the `Model` class and updated the `burgers.py` demo? Since the `Model` class is now separate from `Solver`, I added `time_partition`, `fields`, `function_spaces`, and `meshes` to the signatures of `get_` functions.
> 
> Do you agree with this? I find it a bit messy, but I don't have a different suggestion.

@dundo just a few quick observations:
 - The Model class itself is still appealing partially from a separation of concerns and, as @jwallwork23 noted last week, it also somewhat isolates the user functions related to the problem definition independent of the time or space discretisation. But, as you have noticed, this makes the variables passed to the methods messy.
 - Some ideas (though I don't think they make it any less messy):
     - if we added `get_field_names` to the `Model` and move `field_types` and `steady` to the `Solver` the `TimePartition` class would be just related to the time discrimination
     - `get_initial_condition` like `get_function_space` takes `mesh` and `function_space` in the arguments; or just `mesh` and/or internally calls `get_function_spaces(mesh)`
     -  if you strip the arguments for `get_solver` back to those independent of `Solver`, I think you get: `mesh`,`t_start`,`t_end`,`dt`,`initial_condition`; - this is still messy but potentially independent from the time and space discritisation used.
     - if you strip the arguments for `get_qoi` back you get: `mesh`,`dt`,`solution` or `solutions`

Your unstated question is a good one: whether the benefit of making the user input mostly independent is outweighed by added complexity to the API?

ddundo commented:
Thanks @acse-ej321!
>   * if we added `get_field_names` to the `Model` and move `field_types` and `steady` to the `Solver` the `TimePartition` class would be just related to the time discrimination

I agree but I don't want to make changes related to #181 in this PR since this will require more discussion I think.

>   * `get_initial_condition` like `get_function_space` takes `mesh` and `function_space` in the arguments; or just `mesh` and/or internally calls `get_function_spaces(mesh)`

`Solver` already builds and stores function spaces (in `Solver._fs`) so I don't want to call `get_function_spaces` again since that will rebuild them. 

>   * if you strip the arguments for `get_solver` back to those independent of `Solver`, I think you get: `mesh`,`t_start`,`t_end`,`dt`,`initial_condition`; - this is still messy but potentially independent from the time and space discritisation used.

So you would prefer not to pass a `TimePartition` instance as an argument? I don't see a problem with doing that, to be honest :)

>   * if you strip the arguments for `get_qoi` back you get: `mesh`,`dt`,`solution` or `solutions`

I would prefer to keep arguments the same in all methods and to pass everything
1. for simplicity
2. and because someone might actually use them. For example, we currently don't use `Solver.function_spaces` in `get_qoi` in any of our demos, but we definitely might want to

ddundo committed:
Merge branch 'develop' into 238_solver_class

ddundo committed:
Add .cache/ to .gitignore

ddundo committed:
Linting

ddundo committed:
Updates with fields

ddundo committed:
Updates after introducing Field class

ddundo committed:
`test_model.py`: expand notimplemented error test

ddundo committed:
Updates after introducing Field class again

ddundo committed:
Update `test_error_estimation.py`

ddundo committed:
Update some demos

ddundo committed:
WIP `burgers_time_integrated.py` (to discuss)

ddundo commented on demos/burgers_time_integrated.py:
```diff
@@ -33,13 +33,21 @@ def get_initial_condition(mesh_seq):
-def get_solver(mesh_seq):
-    def solver(index):
-        u, u_ = mesh_seq.field_functions["u"]
+class BurgersModel(Model):
+    def get_initial_condition(self, time_partition, meshes, fields, function_spaces):
+        fs = function_spaces["u"][0]
+        x, y = SpatialCoordinate(meshes[0])
+        return {"u": Function(fs).interpolate(as_vector([sin(pi * x), 0]))}
+
+    def get_solver(
+        self, index, time_partition, meshes, field_functions, function_spaces
+    ):
```
We have a `Model` class which is separate from the `Solver` class, so we need to pass `time_partition`, `meshes`, `field_functions`, and `function_spaces` from the `Solver` to `Model`'s `get_*` methods.

jwallwork23 replied:
I'd prefer not to keep passing all these arguments repeatedly. Could we not pass the `time_partition` to the `__init__` method of the `Model`? Also, we should be able to determine the fields and function spaces using that information, no? Then we'd only need to pass meshes and indices to these methods.

ddundo commented on demos/burgers_time_integrated.py:
```diff
@@ -51,19 +59,68 @@ def solver(index):
         )
 
         # Time integrate from t_start to t_end
-        t_start, t_end = mesh_seq.subintervals[index]
-        dt = mesh_seq.time_partition.timesteps[index]
+        t_start, t_end = time_partition.subintervals[index]
+        dt = time_partition.timesteps[index]
         t = t_start
-        qoi = mesh_seq.get_qoi(index)
+        qoi = self.get_qoi(
+            index, time_partition, meshes, field_functions, function_spaces
+        )
         while t < t_end - 1.0e-05:
             solve(F == 0, u, ad_block_tag="u")
-            mesh_seq.J += qoi(t)
             yield
+            self.J += qoi(t)
```
Time-integrated qoi requires us to communicate `J` from the `Model` to the `Solver`.

Similarly for goal-oriented error estimation, where we currently use `GoalOrientedMeshSeq.read_forms({"field": Form})`. We need to modify it so that `{"field": Form}` gets passed from `Model` to `Solver`

ddundo replied:
Simplest solution would be to pass `J` as an argument to `get_*` functions as well. I think this is fine since we plan to add QoIs for Hessian-based adaptations too (#280)

jwallwork23 replied:
I think the way we handle the `J` attribute needs some rethinking anyway. How about you draft something and then we can comment further on that?

ddundo committed:
Remove `Model` class (WIP)

ddundo committed:
Update some demos

ddundo committed:
Update test_utils.py

ddundo committed:
Fixes

ddundo committed:
Fix adjoint/test_adjoint.py

ddundo committed:
Fix test_adjoint_mesh_seq apart from p-enrichment

ddundo committed:
More partial fixes

ddundo committed:
Fix test_adjoint_mesh_seq.py

ddundo committed:
Fix test_fp_iteration.py

ddundo committed:
(WIP) update demos; use annotate_qoi decorator

joewallwork commented:
Ah, this PR was automatically closed when I deleted the `develop` branch and I'm unable to reopen it. The branch still exists, though.
